### PR TITLE
Razor boolean attribute special case handled correctly

### PR DESCRIPTION
**Pull request by ajaybhargavb**
Issue - #2769 
@NTaylorMullen @dougbu

### CODE REVIEW

**Comment by dougbu** [comment]
@ajaybhargavb is this the issue or PR you and @NTaylorMullen discussed a lot today? If yes, does it currently match any decisions from those discussions? Basically is it too early to review?

**Comment by ajaybhargavb** [comment]
@dougbu, yes. Don't review now. I'll send an update.

**Comment by ajaybhargavb** [comment]
Updated. @dougbu @NTaylorMullen

File: src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs [hunk 1]
```diff
@@ -529,85 +527,52 @@ public virtual void WriteLiteralTo([NotNull] TextWriter writer, string value)
...
-                        WritePositionTaggedLiteral(writer, prefix);
-                        first = false;
-                    }
+                        string stringValue = null;
 
-                    if (!string.IsNullOrEmpty(attrVal.Prefix))
-                    {
-                        WritePositionTaggedLiteral(writer, attrVal.Prefix);
-                    }
+                        stringValue = positionTaggedAttributeValue.Value as string;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: combine w/ previous line: `var stringValue = positionTaggedAttributeValue.Value as string;`

File: src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs [hunk 2]
```diff
@@ -529,85 +527,52 @@ public virtual void WriteLiteralTo([NotNull] TextWriter writer, string value)
...
 
-                    // Intentionally using is+cast here for performance reasons. This is more performant than as+bool?
-                    // because of boxing.
-                    if (val.Value is bool)
+                    // The value is just the bool 'true', write the attribute name instead of the string 'true'.
+                    WriteAttributeValue(writer, attributeValue, name, suffix.Position);
+                }
+                else
+                {
+                    WritePositionTaggedLiteral(writer, prefix);
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: Shouldn't `prefix` be written out only if the first value is not `@null` or equivalent?

Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: The decision was to not write prefix only if the value contains `@false` or `@null` with no surrounding whitespace.

File: src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs [hunk 3]
```diff
@@ -621,6 +586,42 @@ public virtual string Href([NotNull] string contentPath)
...
+        {
+            var positionTaggedAttributeValue = attributeValue.Value;
+
+            if (!string.IsNullOrEmpty(attributeValue.Prefix))
+            {
+                WritePositionTaggedLiteral(writer, attributeValue.Prefix);
+            }
+
+            BeginContext(attributeValue.Value.Position, sourceLength, isLiteral: attributeValue.Literal);
+            // The extra branching here is to ensure that we call the Write*To(string) overload where
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: Add a blank line above to make this grouping even more clear.

File: src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs [hunk 4]
```diff
@@ -529,85 +527,52 @@ public virtual void WriteLiteralTo([NotNull] TextWriter writer, string value)
             }
             else
             {
-                for (var i = 0; i < values.Length; i++)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: Suggest a comment here and in the commit description that informs reader of the two rules now enforced / implemented in this large block i.e. `null` / `bool` handling when there's just one value and `null` plus prefix handling in all other cases.

**Comment by dougbu** [comment]
:watch: mostly for description wording

File: src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs [hunk 5]
```diff
@@ -529,85 +527,52 @@ public virtual void WriteLiteralTo([NotNull] TextWriter writer, string value)
...
+                if (values.Length == 1 && values[0].Prefix == "" &&
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]:
```C#
var attributeValue = values.SingleOrDefault();
if (attributeValue?.Prefix == "" &&
    (attributeValue.Value.Value is bool || attributeValue.Value.Value == null))
{
    var positionTaggedAttributeValue = attributeValue.Value;
```

File: src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs [hunk 6]
```diff
@@ -529,85 +527,52 @@ public virtual void WriteLiteralTo([NotNull] TextWriter writer, string value)
...
+                    for (var i = 0; i < values.Length; i++)
                     {
-                        if ((bool)val.Value)
-                        {
-                            stringValue = name;
-                        }
-                        else
+                        var attributeValue = values[i];
+                        var positionTaggedAttributeValue = attributeValue.Value;
+                        var next = i == values.Length - 1 ?
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: Move down to where it's used

File: src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs [hunk 7]
```diff
@@ -529,85 +527,52 @@ public virtual void WriteLiteralTo([NotNull] TextWriter writer, string value)
...
-                    // value might be a bool. If the value is the bool 'true' we want to write the attribute name
-                    // instead of the string 'true'. If the value is the bool 'false' we don't want to write anything.
-                    // Otherwise the value is another object (perhaps an HtmlString) and we'll ask it to format itself.
-                    string stringValue;
+                    WritePositionTaggedLiteral(writer, prefix);
 
-                    // Intentionally using is+cast here for performance reasons. This is more performant than as+bool?
-                    // because of boxing.
-                    if (val.Value is bool)
+                    // The value is just the bool 'true', write the attribute name instead of the string 'true'.
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: `Instead of the string 'True'`

File: src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs [hunk 8]
```diff
@@ -529,85 +527,52 @@ public virtual void WriteLiteralTo([NotNull] TextWriter writer, string value)
...
+                    WriteAttributeValue(writer, attributeValue, name, suffix.Position);
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: The `SourceLength` is wrong here. It should be `suffix.Position - attributeValue.Value.Position`.

Should add unit tests to validate we're beginning contexts with the correct info.

File: test/Microsoft.AspNet.Mvc.FunctionalTests/HtmlGenerationTest.cs [hunk 9]
```diff
@@ -575,5 +575,22 @@ public class HtmlGenerationTest
...
+            var expectedContent =
+                await ResourceFile.ReadResourceAsync(_resourcesAssembly, outputFile, sourceFile: false);
+
+            // Act
+            var response = await client.GetStringAsync("http://localhost/HtmlGeneration_Home/AttributesWithBooleanValues");
+
+            // Assert
+            Assert.Equal(expectedContent, response);
+        }
     }
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/HtmlGenerationTest.cs]: Should add a test variation that encodes the same file with the `TestHtmlEncoder`. Will make sure the new `WriteAttributeValue` method encodes correctly in the new cases.

**Comment by NTaylorMullen** [comment]
:watch:

File: src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs (hunk 5 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: @NTaylorMullen `SingleOrDefault()` will `throw` if there are more than one elements. Definitely not what we want here. (Further @ajaybhargavb's code is pretty clear and we've got an array so it's already efficient.)

File: test/Microsoft.AspNet.Mvc.FunctionalTests/HtmlGenerationTest.cs (hunk 9 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/HtmlGenerationTest.cs]: +1 it's rather scary that no existing tests covered the complex rules we're removing here. Should have some / all of the variations added to this functional test covered in the `RazorPageTest` class.

File: src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs (hunk 5 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: Dur dur dur. I always mess up singleordefault :+1:

Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: Can combine this if as an else if above then

**Comment by ajaybhargavb** [comment]
Updated.

File: test/Microsoft.AspNet.Mvc.FunctionalTests/HtmlGenerationTest.cs [hunk 10]
```diff
@@ -110,13 +112,14 @@ public class HtmlGenerationTest
...
-        [InlineData("OrderUsingHtmlHelpers", "/HtmlGeneration_Order/Submit")]
-        [InlineData("Product", null)]
-        [InlineData("Script", null)]
+        //[InlineData("EditWarehouse", null)]
+        //[InlineData("Index", null)]
+        //[InlineData("Link", null)]
+        //[InlineData("Order", "/HtmlGeneration_Order/Submit")]
+        //[InlineData("OrderUsingHtmlHelpers", "/HtmlGeneration_Order/Submit")]
+        //[InlineData("Product", null)]
+        //[InlineData("Script", null)]
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/HtmlGenerationTest.cs]: Remove, don't comment out code

Comment (ajaybhargavb) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/HtmlGenerationTest.cs]: Oops. Forgot to uncomment them.

**Comment by NTaylorMullen** [comment]
:shipit: after uncomments.

**Comment by dougbu** [comment]
:shipit:

**Pull request merged**